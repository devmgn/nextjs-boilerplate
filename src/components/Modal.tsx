'use client';

import { useRef, useMemo, useCallback } from 'react';
import {
  disableBodyScroll,
  enableBodyScroll,
  clearAllBodyScrollLocks,
} from 'body-scroll-lock-upgrade';
import FocusLock from 'react-focus-lock';
import { CSSTransition } from 'react-transition-group';
import { useUnmount } from 'react-use';
import styled, { useTheme } from 'styled-components';
import { CSS_TRANSITION_CLASSNAMES } from '@/config/cssTransition';
import ModalProvider from '@/providers/ModalProvider';
import createShouldForwardProp from '@/utils/createShouldForwardProp';
import Backdrop from './Backdrop';
import Portal from './Portal';
import type { WithChildrenProps } from '@/types';
import type { ReactFocusLockProps } from 'react-focus-lock/interfaces';

const StyledModal = styled.div.withConfig({
  shouldForwardProp: createShouldForwardProp('transitionDuration'),
})<NonNullable<Pick<ModalProps, 'transitionDuration'>>>`
  position: relative;
  z-index: ${({ theme }) => theme.zIndex.modal};
  & {
    .Backdrop {
      opacity: 0;
      transition: ${({ theme, transitionDuration }) =>
        theme.transitions.create(['opacity'], { duration: transitionDuration })};
    }
    .ModalRoot {
      opacity: 0;
      transition: ${({ theme, transitionDuration }) =>
        theme.transitions.create(['opacity', 'translate'], { duration: transitionDuration })};
      translate: 0 24px;
    }
  }

  &.enterDone {
    .Backdrop {
      opacity: 1;
    }
    .ModalRoot {
      opacity: 1;
      translate: none;
    }
  }
`;

const ModalRoot = styled.div`
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 24px;
  pointer-events: none;
  outline: none;
  > * {
    pointer-events: auto;
  }
`;

ModalRoot.defaultProps = {
  className: 'ModalRoot',
};

type ModalProps = {
  isOpen: boolean;
  close: () => void;
  transitionDuration?: number;
  closeTriggerAttribute?: string;
  disableCloseOnEscKeydown?: boolean;
  focusLockProps?: Omit<ReactFocusLockProps, 'as' | 'lockProps'>;
} & WithChildrenProps;

const Modal: React.FC<ModalProps> = ({
  isOpen,
  close,
  transitionDuration,
  closeTriggerAttribute = '[data-close-modal="true"]',
  disableCloseOnEscKeydown = false,
  focusLockProps,
  children,
}) => {
  const theme = useTheme();
  const modalRef = useRef<HTMLDivElement>(null);

  const timeout = useMemo(() => {
    return transitionDuration !== undefined
      ? transitionDuration
      : theme.transitions.duration.enteringScreen;
  }, [theme.transitions.duration.enteringScreen, transitionDuration]);

  const closeOnClick = useCallback(
    ({ target }: React.MouseEvent<HTMLDivElement>) => {
      const shouldClose =
        target instanceof HTMLElement
          ? Boolean(target.closest(closeTriggerAttribute)) &&
            modalRef.current?.contains(document.activeElement)
          : false;

      if (shouldClose) {
        close();
      }
    },
    [close, closeTriggerAttribute],
  );

  const closeOnEscKeydown = useCallback(
    ({ key }: React.KeyboardEvent<HTMLDivElement>) => {
      const shouldClose = key === 'Escape' && modalRef.current?.contains(document.activeElement);
      if (disableCloseOnEscKeydown || !shouldClose) {
        return;
      }
      close();
    },
    [close, disableCloseOnEscKeydown],
  );

  const modalRefCallback = useCallback((callback: (element: HTMLDivElement) => void) => {
    if (!modalRef.current) {
      return;
    }
    callback(modalRef.current);
  }, []);

  const lockScroll = useCallback(() => {
    modalRefCallback((element) => {
      disableBodyScroll(element, {
        reserveScrollBarGap: true,
        allowTouchMove: (target) => (target instanceof Node ? element.contains(target) : false),
      });
    });
  }, [modalRefCallback]);

  const clearLockedScroll = useCallback(() => {
    modalRefCallback(enableBodyScroll);
  }, [modalRefCallback]);

  useUnmount(() => {
    clearAllBodyScrollLocks();
    return null;
  });

  return (
    <Portal>
      <CSSTransition
        in={isOpen}
        unmountOnExit
        timeout={{ exit: timeout }}
        classNames={CSS_TRANSITION_CLASSNAMES}
        onEnter={lockScroll}
        onExited={clearLockedScroll}
        nodeRef={modalRef}
      >
        <StyledModal
          ref={modalRef}
          transitionDuration={timeout}
          onClick={closeOnClick}
          onKeyDown={closeOnEscKeydown}
        >
          <Backdrop onClick={close} />
          <FocusLock returnFocus {...focusLockProps} as={ModalRoot} lockProps={{ tabIndex: 0 }}>
            <ModalProvider isOpen={isOpen} close={close}>
              {children}
            </ModalProvider>
          </FocusLock>
        </StyledModal>
      </CSSTransition>
    </Portal>
  );
};

export default Modal;
