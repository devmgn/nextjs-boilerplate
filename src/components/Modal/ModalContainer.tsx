import styled from 'styled-components';

const ModalContainer = styled.div.attrs({ className: 'ModalContainer' })`
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 24px;
  pointer-events: none;
  outline: none;
  > * {
    pointer-events: auto;
  }
`;

export default ModalContainer;
