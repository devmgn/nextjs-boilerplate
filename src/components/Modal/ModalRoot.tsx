import styled from 'styled-components';
import createShouldForwardProp from '@/utils/createShouldForwardProp';
import type { ModalProps } from '.';

const ModalRoot = styled.div.withConfig({
  shouldForwardProp: createShouldForwardProp('transitionDuration'),
})<NonNullable<Pick<ModalProps, 'transitionDuration'>>>`
  position: relative;
  z-index: ${({ theme }) => theme.zIndex.modal};
  & {
    .Backdrop {
      opacity: 0;
      transition: ${({ theme, transitionDuration }) =>
        theme.transitions.create(['opacity'], { duration: transitionDuration })};
    }
    .ModalContainer {
      opacity: 0;
      transition: ${({ theme, transitionDuration }) =>
        theme.transitions.create(['opacity', 'translate'], { duration: transitionDuration })};
      translate: 0 24px;
    }
  }

  &.enter,
  &.enter-done {
    .Backdrop {
      opacity: 1;
    }
    .ModalContainer {
      opacity: 1;
      translate: none;
    }
  }
`;

export default ModalRoot;
