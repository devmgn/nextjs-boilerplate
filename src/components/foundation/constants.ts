export const FONT_FAMILY = {
  SANS_SERIF: 'system-ui, -apple-system, sans-serif',
  SERIF: "serif",
  MONOSPACE: 'monospace, monospace',
};

export const MIN_DEVICE_WIDTH = 320;
