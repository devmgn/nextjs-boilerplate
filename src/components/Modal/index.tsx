'use client';

import { useCallback, useMemo, useRef } from 'react';
import {
  clearAllBodyScrollLocks,
  disableBodyScroll,
  enableBodyScroll,
} from 'body-scroll-lock-upgrade';
import FocusLock from 'react-focus-lock';
import { CSSTransition } from 'react-transition-group';
import { useUnmount } from 'react-use';
import { useTheme } from 'styled-components';
import Backdrop from '../Backdrop';
import Portal from '../Portal';
import ModalContainer from './ModalContainer';
import ModalRoot from './ModalRoot';
import type { WithChildrenProps } from '@/types';
import type { ReactFocusLockProps } from 'react-focus-lock/interfaces';

export type ModalProps = {
  isOpen: boolean;
  close: () => void;
  transitionDuration?: number;
  closeTriggerAttribute?: string;
  disableCloseOnEscKeydown?: boolean;
  focusLockProps?: Omit<ReactFocusLockProps, 'as' | 'lockProps'>;
} & WithChildrenProps;

const Modal: React.FC<ModalProps> = ({
  isOpen,
  close,
  transitionDuration,
  disableCloseOnEscKeydown = false,
  focusLockProps,
  children,
}) => {
  const theme = useTheme();
  const modalRef = useRef<HTMLDivElement>(null);

  const timeout = useMemo(() => {
    return transitionDuration ?? theme.transitions.duration.enteringScreen;
  }, [theme.transitions.duration.enteringScreen, transitionDuration]);

  const closeOnEscKeydown = useCallback(
    ({ key }: React.KeyboardEvent<HTMLDivElement>) => {
      const shouldClose =
        key === 'Escape' && modalRef.current?.contains(document.activeElement);
      if (!disableCloseOnEscKeydown && shouldClose) {
        close();
      }
    },
    [close, disableCloseOnEscKeydown],
  );

  const modalRefCallback = useCallback(
    (callback: (element: HTMLDivElement) => void) => {
      if (!modalRef.current) {
        return;
      }
      callback(modalRef.current);
    },
    [],
  );

  const lockScroll = useCallback(() => {
    modalRefCallback((element) => {
      disableBodyScroll(element, {
        reserveScrollBarGap: true,
        allowTouchMove: (target) =>
          target instanceof Node ? element.contains(target) : false,
      });
    });
  }, [modalRefCallback]);

  const clearLockedScroll = useCallback(() => {
    modalRefCallback(enableBodyScroll);
  }, [modalRefCallback]);

  useUnmount(() => {
    clearAllBodyScrollLocks();
    return null;
  });

  return (
    <Portal>
      <CSSTransition
        in={isOpen}
        unmountOnExit
        timeout={timeout}
        onEnter={lockScroll}
        onExited={clearLockedScroll}
      >
        <ModalRoot
          ref={modalRef}
          transitionDuration={timeout}
          onKeyDown={closeOnEscKeydown}
        >
          <Backdrop onClick={close} />
          <FocusLock
            returnFocus
            {...focusLockProps}
            as={ModalContainer}
            lockProps={{ tabIndex: 0 }}
          >
            {children}
          </FocusLock>
        </ModalRoot>
      </CSSTransition>
    </Portal>
  );
};

export default Modal;
