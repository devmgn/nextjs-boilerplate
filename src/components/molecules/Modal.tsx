import { createContext, useCallback } from 'react';
import { AnimatePresence, m } from 'framer-motion';
import { rgba } from 'polished';
import FocusLockUI from 'react-focus-lock/UI';
import { disablePageScroll, enablePageScroll } from 'scroll-lock';
import styled from 'styled-components';
import { sidecar } from 'use-sidecar';
import { MODAL_TRANSITION_DURATION } from '@/constants';
import Portal from '@/layouts/Portal';

export { default as useModalState } from '@/hooks/useModalState';

const FocusLockSidecar = sidecar(() => import(/* webpackPrefetch: true */ 'react-focus-lock/sidecar'));

export const StyledModal = styled(m.div)`
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 16px;
  background-color: ${rgba('#000', 0.6)};
`;

type ModalContext = {
  isActive: boolean;
  deactivate: () => void;
};

export const modalContext = createContext<ModalContext>({ isActive: false, deactivate: () => {} });

type ModalProps = {
  isActive: boolean;
  deactivate: () => void;
  onDeactivate: (event: React.MouseEvent<HTMLElement>) => void;
  children: React.ReactNode;
};

const Modal: React.FC<ModalProps> = ({ isActive, deactivate, onDeactivate, children }) => {
  const onExitComplete = useCallback(() => {
    enablePageScroll();
  }, []);

  const onAnimationStart = useCallback<(definition: any) => void>((definition) => {
    if ('opacity' in definition && definition.opacity === 1) {
      disablePageScroll();
    }
  }, []);

  return (
    <modalContext.Provider value={{ isActive, deactivate }}>
      <Portal>
        <AnimatePresence mode="wait" onExitComplete={onExitComplete}>
          {isActive && (
            <div>
              <FocusLockUI returnFocus sideCar={FocusLockSidecar}>
                <StyledModal
                  key="modal"
                  initial={{ opacity: 0 }}
                  animate={{ opacity: 1 }}
                  exit={{ opacity: 0 }}
                  transition={{ duration: MODAL_TRANSITION_DURATION }}
                  onClick={onDeactivate}
                  onAnimationStart={onAnimationStart}
                >
                  {children}
                </StyledModal>
              </FocusLockUI>
            </div>
          )}
        </AnimatePresence>
      </Portal>
    </modalContext.Provider>
  );
};

export default Modal;
