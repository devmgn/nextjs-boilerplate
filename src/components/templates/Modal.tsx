import { useCallback, useMemo } from 'react';
import { AnimatePresence, m } from 'framer-motion';
import { rgba } from 'polished';
import ReactFocusLock from 'react-focus-lock';
import { disablePageScroll, enablePageScroll } from 'scroll-lock';
import styled from 'styled-components';
import { TRANSITION } from '@/constants/framerMotion';
import { modalContext } from '@/contexts/modalContext';
import useModal from '@/hooks/useModal';
import Portal from './Portal';

const StyledModal = styled.div`
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 16px;
`;

const Backdrop = styled(m.div)`
  position: absolute;
  inset: 0;
  z-index: -1;
  background-color: ${rgba('#000', 0.6)};
`;

type ModalProps = {
  children: React.ReactNode;
} & Omit<ReturnType<typeof useModal>, 'activate'>;

const Modal: React.FC<ModalProps> = ({ isActive, deactivate, handleDeactivate, children }) => {
  const onExitComplete = useCallback(() => {
    enablePageScroll();
  }, []);

  const onAnimationStart = useCallback<(definition: 'active' | 'inactive') => void>(
    (definition) => {
      if (definition === 'active') {
        disablePageScroll();
      }
    },
    []
  );

  const contextValue = useMemo(() => ({ isActive, deactivate }), [deactivate, isActive]);

  return (
    <modalContext.Provider value={contextValue}>
      <Portal>
        <AnimatePresence mode="wait" onExitComplete={onExitComplete}>
          {isActive && (
            <div tabIndex={-1}>
              <ReactFocusLock returnFocus>
                <StyledModal onClick={handleDeactivate}>
                  <Backdrop
                    key="modal-backdrop"
                    initial="inactive"
                    animate="active"
                    exit="inactive"
                    variants={{
                      active: { opacity: 1 },
                      inactive: { opacity: 0 },
                    }}
                    transition={TRANSITION}
                    onAnimationStart={onAnimationStart}
                    role="presentation"
                    data-modal-deactivate
                  />
                  {children}
                </StyledModal>
              </ReactFocusLock>
            </div>
          )}
        </AnimatePresence>
      </Portal>
    </modalContext.Provider>
  );
};

export default Modal;
