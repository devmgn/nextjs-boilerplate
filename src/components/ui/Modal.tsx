import React, { createContext, useCallback, useEffect, useState } from 'react';
import FocusLock from 'react-focus-lock';
import { disablePageScroll, enablePageScroll } from 'scroll-lock';
import styled from 'styled-components';
import { fadeIn, fadeOut } from '@/components/foundation/keyframes';
import Portal from '@/components/ui/Portal';

const ModalDiv = styled.div`
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 16px;
  background-color: rgba(0, 0, 0, 0.6);
  animation: ${fadeIn} 0.3s ease-in-out forwards;

  &[aria-hidden='true'] {
    animation-name: ${fadeOut};
  }
  &[aria-busy='true'] {
    pointer-events: none;
  }
`;

type ModalContext = {
  setIsHidden: React.Dispatch<React.SetStateAction<boolean>>;
};

export const modalContext = createContext<ModalContext>({} as ModalContext);

type ModalProps = {
  isActive: boolean;
  setIsActive: React.Dispatch<React.SetStateAction<boolean>>;
  children: React.ReactNode;
};

const Modal: React.FC<ModalProps> = ({ isActive, setIsActive, children }) => {
  const [isHidden, setIsHidden] = useState(false);

  const onClick = useCallback(({ target }: React.MouseEvent | React.TouchEvent) => {
    if (target === document.querySelector(ModalDiv)) {
      setIsHidden(true);
    }
  }, []);

  const onAnimationStart = useCallback(() => {
    if (isHidden) {
      return;
    }
    disablePageScroll();
  }, [isHidden]);

  const onAnimationEnd = useCallback(() => {
    if (isHidden) {
      setIsActive(false);
      enablePageScroll();
    }
  }, [isHidden, setIsActive]);

  useEffect(() => {
    const onEscKeyDown = ({ key }: KeyboardEvent) => {
      if (key === 'Escape') {
        setIsHidden(true);
      }
    };

    if (isActive) {
      document.addEventListener('keydown', onEscKeyDown);
    }
    return () => document.removeEventListener('keydown', onEscKeyDown);
  }, [isActive]);

  return (
    <>
      {isActive && (
        <modalContext.Provider value={{ setIsHidden }}>
          <Portal>
            <FocusLock autoFocus={false}>
              <ModalDiv
                aria-hidden={isHidden}
                onClick={onClick}
                onAnimationStart={onAnimationStart}
                onAnimationEnd={onAnimationEnd}
              >
                {children}
              </ModalDiv>
            </FocusLock>
          </Portal>
        </modalContext.Provider>
      )}
    </>
  );
};

export default Modal;
