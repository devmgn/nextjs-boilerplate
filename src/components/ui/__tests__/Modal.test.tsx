import { render } from '@testing-library/react';
import { ROOT_CONTAINER_ID } from '@/constants';
import Modal from '../Modal';

const defaultState = {
  isActive: false,
  deactivate: jest.fn(),
  onDeactivate: jest.fn(),
};

const activeState = { ...defaultState, isActive: true };

describe('Modal', () => {
  let modal: HTMLElement | null;

  beforeEach(() => {
    modal = document.createElement('div');
    modal.id = ROOT_CONTAINER_ID;
    document.body.appendChild(modal);
  });

  afterEach(() => (modal = null));

  describe('スナップショットテスト', () => {
    test('初期表示', () => {
      const { asFragment } = render(<Modal {...defaultState}>children</Modal>, { container: modal! });
      expect(asFragment()).toMatchSnapshot();
    });

    test('isActiveがtrueの時', () => {
      const { asFragment } = render(<Modal {...activeState}>children</Modal>, { container: modal! });
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
