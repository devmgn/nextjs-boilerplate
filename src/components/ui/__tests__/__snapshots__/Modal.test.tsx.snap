// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal スナップショットテスト isActiveがtrueの時 1`] = `
<DocumentFragment>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  .c0 {
  position: fixed;
  inset: 0;
  display: grid;
  place-items: center;
  padding: 16px;
  background-color: rgba(0,0,0,0.6);
}

<div
    data-focus-lock-disabled="false"
  >
    <div
      class="c0"
      style="opacity: 0;"
    >
      children
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</DocumentFragment>
`;

exports[`Modal スナップショットテスト 初期表示 1`] = `<DocumentFragment />`;
